README.md
# washhub - an experimental external Github plugin for [Wash](https://puppetlabs.github.io/wash/)

## TL;DR

`washhub` integrates your Github content into your [Wash](https://puppetlabs.github.io/wash/) filesystem.
You can navigate your organisations, repos, directories and files.

## Installation

1. Of course you already have Wash installed and working on your local system. If not, see [Wash Docs](https://puppetlabs.github.io/wash/) for instructions on how to get up and running!
1. Make sure you have Go installed on your system (tested with go 1.12.5)
1. Clone the `washhub` repo and build the plugin. The executable name needs to be `github`:

    ```bash
    git clone https://github.com/timidri/washhub
    cd washhub
    go build -o github
    ```

1. Configure your github token in `~/.washhub.yaml`:

    ```bash
    cat > ~/.washhub.yaml
    github_token: <my_github_token>
    ```

    Alternatively, you can use user credentials:

    ```bash
    cat > ~/.washhub.yaml
    github_user: <my_github_user>
    github_password: <my_github_password>
    ```

1. Configure Wash to use the washhub plugin by editing the `wash.yaml` configuration file (`~/.puppetlabs/wash/wash.yaml` by default):

    ```yaml
    external-plugins:
      - script: '/path/to/washhub/github'
    ```

1. Enjoy!

## How to use

Washhub supports the following Wash actions:

* `list`: organizations, repos and directories inside repos
* `read`: files

## Known bugs

Listing orgs with lots of repos can result in an empty repository listing. When running wash in debug mode (`wash --loglevel debug`) the following error is displayed:

```quote
DEBU FUSE: Find .git in /github/<org> errored: script returned a non-zero exit code of 1. stderr output: Error: GET https://api.github.com/orgs/<org>/repos?per_page=99: 502 Server Error []
```

This erroneous behaviour is documented [here](https://github.com/google/go-github/issues/999) and I didn't find a work-around for now. Ideas welcome!

## Limitations of `washhub`

* Supported are repos, directories and files, other objects such as projects, issues or commits are not supported
* No `metadata` support yet
* `stream` or `exec` not supported because Github doesn't support that
* non-authenticated operation not supported
* You can hit Github API limits if you use `washhub` _a lot_
  
## Author

dimitri@puppet.com


CHANGELOG.md
# washhub - CHANGELOG

## 2019-07-18

* Changed plugin name to `github` for clarity

## 2019-06-12

* Removed the plugin name from init's response, wash now requires the plugin to have the same name as the executable
* Introduced a const to contain plugin name


cmd/init.go
// Copyright © 2019 NAME HERE <EMAIL ADDRESS>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package cmd

import (
	"fmt"

	"github.com/spf13/cobra"
)

// PluginName is the name of our plugin
const PluginName = "github"

// initCmd represents the init command
var initCmd = &cobra.Command{
	Use:   "init",
	Short: "Initialize the plugin",
	Run: func(cmd *cobra.Command, args []string) {
		response := `
		{
			"name": "github",
			"methods": [
				"list"
			],
			"cache_ttls": {
				"list": 600,
				"read": 30
			}		
		}`
		fmt.Println(response)
	},
}

func init() {
	rootCmd.AddCommand(initCmd)
}


cmd/read.go
// Copyright © 2019 NAME HERE <EMAIL ADDRESS>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package cmd

import (
	b64 "encoding/base64"
	"fmt"
	"strings"

	"github.com/spf13/cobra"
)

// readCmd represents the read command
var readCmd = &cobra.Command{
	Use:   "read <path> ''",
	Short: "Read content at <path>",
	Args:  cobra.ExactArgs(2),
	Run: func(cmd *cobra.Command, args []string) {
		path := strings.TrimPrefix(args[0], "/"+PluginName+"/")
		org, tail := SplitPath(path)
		repo, path := SplitPath(tail)
		printFile(org, repo, path)
	},
}

func printFile(org string, repo string, path string) {
	fileContent, _, err := FetchRepositoryContent(org, repo, path)
	HandleError(err)
	decodedContent, _ := b64.StdEncoding.DecodeString(*fileContent.Content)
	fmt.Println(string(decodedContent))
}

func init() {
	rootCmd.AddCommand(readCmd)
}


cmd/root.go
// Copyright © 2019 NAME HERE <EMAIL ADDRESS>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package cmd

import (
	"fmt"
	"os"

	homedir "github.com/mitchellh/go-homedir"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"
)

var cfgFile string

// rootCmd represents the base command when called without any subcommands
var rootCmd = &cobra.Command{
	Use: "washhub",
}

// Execute adds all child commands to the root command and sets flags appropriately.
// This is called by main.main(). It only needs to happen once to the rootCmd.
func Execute() {
	if err := rootCmd.Execute(); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}

func init() {
	cobra.OnInitialize(initConfig)

	// Here you will define your flags and configuration settings.
	// Cobra supports persistent flags, which, if defined here,
	// will be global for your application.
	rootCmd.PersistentFlags().StringVar(&cfgFile, "config", "", "config file (default is $HOME/.washhub.yaml)")

	// Cobra also supports local flags, which will only run
	// when this action is called directly.
	rootCmd.Flags().BoolP("toggle", "t", false, "Help message for toggle")
}

// initConfig reads in config file and ENV variables if set.
func initConfig() {
	if cfgFile != "" {
		// Use config file from the flag.
		viper.SetConfigFile(cfgFile)
	} else {
		// Find home directory.
		home, err := homedir.Dir()
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}

		// Search config in home directory with name ".washhub" (without extension).
		viper.AddConfigPath(home)
		viper.SetConfigName(".washhub")
	}

	viper.AutomaticEnv() // read in environment variables that match

	// If a config file is found, read it in.
	if err := viper.ReadInConfig(); err == nil {
		// fmt.Println("Using config file:", viper.ConfigFileUsed())
		// fmt.Println("username:", viper.Get("github_user"))
	}
}


cmd/utils.go
package cmd

import (
	"context"
	"fmt"
	"os"
	"strings"

	"golang.org/x/oauth2"

	"github.com/google/go-github/github"
	"github.com/spf13/viper"
)

var githubClient *github.Client

// GithubClient returns a github client as a singleton
func GithubClient() *github.Client {
	if githubClient == nil {
		if token() != "" {
			ts := oauth2.StaticTokenSource(
				&oauth2.Token{AccessToken: token()},
			)
			tc := oauth2.NewClient(context.Background(), ts)
			githubClient = github.NewClient(tc)
		} else if UserName() == "" || password() == "" {
			fmt.Fprintf(os.Stderr, "Please specify github_token or github_user and github_password in ~/.washhub.yaml")
			os.Exit(1)
		} else {
			tp := github.BasicAuthTransport{
				Username: UserName(),
				Password: password(),
			}
			githubClient = github.NewClient(tp.Client())
		}
	}

	return githubClient
}

// UserName returns the configured user name
func UserName() string {
	user := viper.Get("github_user")
	if user == nil {
		return ""
	}
	return user.(string)
}

func password() string {
	password := viper.Get("github_password")
	if password == nil {
		return ""
	}
	return password.(string)
}

func token() string {
	token := viper.Get("github_token")
	if token == nil {
		return ""
	}
	return token.(string)
}

// FetchRepositoryContent returns the content at path (a dir listing or file content)
func FetchRepositoryContent(username string, repo string, path string) (*github.RepositoryContent, []*github.RepositoryContent, error) {
	fileContent, dirContent, _, err := GithubClient().Repositories.GetContents(context.Background(), username, repo, path, nil)
	return fileContent, dirContent, err
}

// SplitPath splits a path in a head part and a tail part
// examples:
//   SplitPath("a") -> ("a","")
//   SplitPath("a/b") -> ("a","b")
//   SplitPath("a/b/c") => ("a", "b/c")
func SplitPath(path string) (string, string) {
	parts := strings.SplitN(path, string(os.PathSeparator), 2)
	head := parts[0]
	tail := ""
	if len(parts) > 1 {
		tail = parts[1]
	}
	return head, tail
}

// HandleError provides trivial error handling
func HandleError(err error) {
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v", err.Error())
		os.Exit(1)
	}
}


cmd/list.go
// Copyright © 2019 NAME HERE <EMAIL ADDRESS>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package cmd

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/google/go-github/github"
	"github.com/spf13/cobra"
	"github.com/tidwall/pretty"
)

// listCmd represents the list command
var listCmd = &cobra.Command{
	Use:   "list <path> ''",
	Short: "List content at <path>",
	Args:  cobra.ExactArgs(2),
	Run: func(cmd *cobra.Command, args []string) {
		path := strings.TrimSuffix(args[0], "/")
		user, _, err := GithubClient().Users.Get(context.Background(), "")
		HandleError(err)

		if path == "/"+PluginName { // we are at top level
			listOrganisations(user)
		} else {
			path = strings.TrimPrefix(path, "/"+PluginName+"/")
			if strings.Contains(path, "/") { // we have at least org and repo in the path
				listContent(path)
			} else { // we have only an org or a user in the path
				if user.GetLogin() == path {
					// list user's repos
					listUserRepos(path)
				} else {
					// list org's repos
					listOrgRepos(path)
				}
			}
		}
		os.Exit(0)
	},
}

type attributes struct {
	Size int `json:"size"`
}

type entry struct {
	Name       string     `json:"name"`
	Methods    []string   `json:"methods"`
	Attributes attributes `json:"attributes"`
}

func listContent(path string) {
	org, tail := SplitPath(path)
	repo, filePath := SplitPath(tail)
	_, dirContent, err := FetchRepositoryContent(org, repo, filePath)
	HandleError(err)
	entries := directoryToEntries(dirContent)
	PrintEntries(entries)
}

func listOrgRepos(org string) {
	opt := &github.RepositoryListByOrgOptions{
		// setting PerPage to 99 instead of 100 to work around
		// https://github.com/google/go-github/issues/999
		// but work-around doesn't seem to work...
		ListOptions: github.ListOptions{PerPage: 99},
	}
	// get all pages of results
	var allRepos []*github.Repository
	for {
		repos, resp, err := GithubClient().Repositories.ListByOrg(context.Background(), org, opt)
		HandleError(err)

		allRepos = append(allRepos, repos...)
		if resp.NextPage == 0 {
			break
		}
		opt.Page = resp.NextPage
	}
	entries := reposToEntries(allRepos)
	PrintEntries(entries)
}

func listUserRepos(userName string) {
	opt := &github.RepositoryListOptions{
		Affiliation: "owner", // we want only owner's repos
		// setting PerPage to 99 instead of 100 to work around
		// https://github.com/google/go-github/issues/999
		// but work-around doesn't seem to work...
		ListOptions: github.ListOptions{PerPage: 99},
	}
	// get all pages of results
	var allRepos []*github.Repository
	for {
		// passing empty string for user name to get repos for authenticated user
		repos, resp, err := GithubClient().Repositories.List(context.Background(), "", opt)
		HandleError(err)

		allRepos = append(allRepos, repos...)
		if resp.NextPage == 0 {
			break
		}
		opt.Page = resp.NextPage
	}
	entries := reposToEntries(allRepos)
	PrintEntries(entries)
}

func listOrganisations(user *github.User) {
	orgs, _, err := GithubClient().Organizations.ListOrgMemberships(context.Background(), nil)
	HandleError(err)

	var entries []*entry

	userEntry := &entry{
		Name:    user.GetLogin(),
		Methods: []string{"list"},
	}

	entries = append(entries, userEntry)

	for _, org := range orgs {
		orgEntry := &entry{
			Name:    *org.GetOrganization().Login,
			Methods: []string{"list"},
		}
		entries = append(entries, orgEntry)
	}
	PrintEntries(entries)
}

// PrintEntries prints the entries
func PrintEntries(entries []*entry) {
	json, _ := json.Marshal(entries)
	fmt.Println(string(pretty.Pretty(json)))
}

// PrintEntry prints one entry
func PrintEntry(entry *entry) {
	json, _ := json.Marshal(entry)
	fmt.Println(string(pretty.Pretty(json)))
}

// directoryToEntries converts github dirContent to entries
func directoryToEntries(dirEntries []*github.RepositoryContent) []*entry {
	entries := make([]*entry, len(dirEntries))
	for i, dirEntry := range dirEntries {
		myEntry := &entry{
			Name: *dirEntry.Name,
			Attributes: attributes{
				Size: *dirEntry.Size,
			},
		}
		if *dirEntry.Type == "file" {
			myEntry.Methods = []string{"read"}
		} else {
			myEntry.Methods = []string{"list"}
		}
		entries[i] = myEntry
	}
	return entries

}

// reposToEntries converts repos to entries
func reposToEntries(repos []*github.Repository) []*entry {
	entries := make([]*entry, len(repos))
	for i, repo := range repos {
		myEntry := &entry{
			Name:    *repo.Name,
			Methods: []string{"list"},
		}
		entries[i] = myEntry
	}
	return entries
}

func init() {
	rootCmd.AddCommand(listCmd)
}
